apps/shell/tash_command.c

```diff
@@ -433,46 +433,71 @@ int check_exclam_cmd(char *buff)
 
 	/* Find the '!' in the input character */
 
-	while ((exclam_ptr = strchr(buff, ASCII_EXCLAM)) != NULL) {
+	int qc = 0;		/* Flag to check whether ! is inside quotation mark */
+	char prev = '\0';	/* Store the previous character for checking escape sequence */
 
-		exclam_nextptr = exclam_ptr + 1;
-		if (*exclam_nextptr == (char)ASCII_SPACE || *exclam_nextptr == (char)ASCII_LF || *exclam_nextptr == (char)ASCII_SEMICOLON || *exclam_nextptr == (char)ASCII_NUL) {
-			/* If the last character in buff is '!', it will be handled as character, not history command. */
+	while (*buff != '\0') {
 
-			buff = exclam_nextptr + 1;
+		if(*buff == '\'' && qc == 0 && prev != '\\') {
```

Comment:
Could you let me know why checked '\\' ?

Reply:
If ! is with single quotes, it will be considered as character. This is to check the start and end of single quotes. It is escape sequence (\'). Without this it was throwing error similar to not end of quote

Reply:
Linux terminal do not supported [ \' ]. (from Ubuntu 18.04)
Could you please check below command?
`echo 'test\'test'`

Reply:
Yes, I checked it. It does not get supported by linux terminal. 
I updated it. Please check it.